proof/refine/ARM/Schedule_R.thy

```diff
@@ -3040,5 +2990,662 @@ lemma ct_released_cross_weak:
+                     simp: obj_at_def is_tcb is_sc_obj
+                     elim: valid_sched_context_size_objsI)
+  apply clarsimp
+  apply (prop_tac "bound (read_tcb_refill_ready t s)")
+   apply (clarsimp intro!: no_ofailD[OF no_ofail_read_tcb_refill_ready]
+                     simp: pred_tcb_at_def obj_at_def)
+  apply (frule state_relation_pspace_relation)
+  apply (clarsimp simp: pspace_relation_def)
+  apply (drule_tac x=t in bspec, blast)
+  apply (drule_tac x="(t, other_obj_relation)" in bspec, clarsimp)
+  apply (clarsimp simp: other_obj_relation_def tcb_relation_def obj_at'_def projectKOs)
+  apply (prop_tac "bound (readTCBRefillReady t s')")
+   apply (clarsimp intro!: no_ofailD[OF no_ofail_readTCBRefillReady])
+   apply (clarsimp simp: obj_at'_def projectKOs)
+   apply (rule_tac x=scp in exI; clarsimp)
+  apply clarsimp
+  apply (clarsimp simp: read_tcb_refill_ready_def readTCBRefillReady_def oassert_opt_def
+                        readRefillReady_simp read_sc_refill_ready_simp refill_ready'_def obj_at'_def
+                        projectKOs
+                 dest!: read_tcb_obj_ref_SomeD read_sched_context_SomeD readSchedContext_SomeD
+                        threadRead_SomeD
+                 split: option.split_asm)
+  apply (rename_tac sc' sc_ptr tcbPtr)
+  apply (prop_tac "valid_sched_context' sc' s' \<and> valid_sched_context_size' sc'")
+   apply (frule_tac k=sc' and p=sc_ptr in sc_ko_at_valid_objs_valid_sc'[rotated])
+    apply (clarsimp simp: obj_at'_def projectKOs)
+   apply simp
+  apply (frule state_relation_pspace_relation)
+  apply (prop_tac "sc_relation sc n sc'")
+   apply (clarsimp simp: pspace_relation_def)
+   apply (drule_tac x=sc_ptr in bspec, blast)
+   apply (fastforce simp: obj_at'_def projectKOs split: if_splits)
+  apply (frule refills_heads_equal_valid_sched_context')
+    apply (clarsimp simp: vs_all_heap_simps valid_refills_def rr_valid_refills_def
+                   split: if_splits)
+   apply (fastforce dest!: sc_ko_at_valid_objs_valid_sc'
+                     simp: obj_at'_def projectKOs)
+  apply (clarsimp simp: kernelWCETTicks_def state_relation_def)
+  done
+
+lemma getReleaseQueue_corres[corres]:
+  "corres (=) \<top> \<top> (gets release_queue) (getReleaseQueue)"
+  unfolding getReleaseQueue_def
+  apply (rule corres_gets_trivial)
+  by (clarsimp simp: state_relation_def release_queue_relation_def)
+
+lemma releaseQNonEmptyAndReady_simp:
+  "releaseQNonEmptyAndReady s
+    = (if ksReleaseQueue s = []
+          then Some False
+          else readTCBRefillReady (head (ksReleaseQueue s)) s)"
+  by (clarsimp simp: releaseQNonEmptyAndReady_def readReleaseQueue_def asks_def obind_def)
+
+lemma releaseQNonEmptyAndReady_eq:
+  "\<lbrakk>(s, s') \<in> state_relation; pspace_aligned s; pspace_distinct s;
+     valid_objs s; valid_release_q s;
+     active_sc_valid_refills s; valid_objs' s'\<rbrakk>
+   \<Longrightarrow> read_release_q_non_empty_and_ready s = releaseQNonEmptyAndReady s'"
```

Comment:
This is like a corres rule for `lookup`, instead of `nondet_monad`. Here the relation between the values is equality but it can be something like `tcb_relation`, I suppose.